metal_header/sifive_i2c0.c++

```diff
@@ -0,0 +1,303 @@
+/* Copyright 2019 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include <sifive_i2c0.h>
+
+sifive_i2c0::sifive_i2c0(std::ostream &os, const fdt &dtb)
+    : Device(os, dtb, "sifive,i2c0") {
+  /* Count the number of I2Cs */
+  num_i2cs = 0;
+  dtb.match(std::regex(compat_string), [&](node n) { num_i2cs += 1; });
+}
+
+void sifive_i2c0::create_defines() {
+  int i = 0;
+  dtb.match(std::regex(compat_string), [&](node n) {
+    uint32_t num_interrupts = n.get_fields_count<uint32_t>("interrupts");
+    emit_def("__METAL_" + n.handle_cap() + "_INTERRUPTS",
+             std::to_string(num_interrupts));
+
+    n.maybe_tuple("baud", tuple_t<uint32_t>(),
+                  [&]() {
+                    emit_def("__METAL_DT_I2C0_BAUDRATE" + std::to_string(i),
+                             std::to_string(0));
+                  },
+                  [&](uint32_t baud) {
+                    emit_def("__METAL_DT_I2C0_BAUDRATE" + std::to_string(i),
+                             std::to_string(baud));
+                  });
+
+    if (n.field_exists("master")) {
+      emit_def("__METAL_DT_I2C0_MODE" + std::to_string(i), "METAL_I2C_MASTER");
+    }
+
+    if (n.field_exists("slave")) {
+      emit_def("__METAL_DT_I2C0_MODE" + std::to_string(i), "METAL_I2C_SLAVE");
+    }
```

Comment:
Let's hold back on defining the speed and mode of the I2C in the devicetree. I'd rather have the application specify it at first before we commit to a devicetree representation of this information.